src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -230,125 +285,150 @@ RouterExecutorRun(QueryDesc *queryDesc, ScanDirection direction, long count, Tas
- * tables. A distributed modification is successful if any placement of the
- * distributed table is successful. ExecuteDistributedModify returns the number
- * of modified rows in that case and errors in all others. This function will
- * also generate warnings for individual placement failures.
+ * ExecuteTaskAndStoreResults executes the task on the remote node, retrieves
+ * the results and stores them in a tuple store. If the task fails on one of
+ * the placements, the function retries it on other placements.
  */
-static int32
-ExecuteDistributedModify(Task *task)
+static bool
+ExecuteTaskAndStoreResults(QueryDesc *queryDesc, Task *task,
+						   bool isModificationQuery,
+						   bool expectResults)
 {
-	int32 affectedTupleCount = -1;
+	TupleDesc tupleDescriptor = queryDesc->tupDesc;
+	EState *executorState = queryDesc->estate;
+	MaterialState *routerState = (MaterialState *) queryDesc->planstate;
+	bool resultsOK = false;
+	List *taskPlacementList = task->taskPlacementList;
 	ListCell *taskPlacementCell = NULL;
 	List *failedPlacementList = NIL;
 	ListCell *failedPlacementCell = NULL;
+	int64 affectedTupleCount = -1;
+	bool gotResults = false;
 
-	foreach(taskPlacementCell, task->taskPlacementList)
+	/*
+	 * Try to run the query to completion on one placement. If the query fails
+	 * attempt the query on the next placement.
+	 */
+	foreach(taskPlacementCell, taskPlacementList)
 	{
 		ShardPlacement *taskPlacement = (ShardPlacement *) lfirst(taskPlacementCell);
 		char *nodeName = taskPlacement->nodeName;
 		int32 nodePort = taskPlacement->nodePort;
+		bool queryOK = false;
+		int64 currentAffectedTupleCount = 0;
+		PGconn *connection = GetOrEstablishConnection(nodeName, nodePort);
 
-		PGconn *connection = NULL;
-		PGresult *result = NULL;
-		char *currentAffectedTupleString = NULL;
-		int32 currentAffectedTupleCount = -1;
-
-		Assert(taskPlacement->shardState == FILE_FINALIZED);
-
-		connection = GetOrEstablishConnection(nodeName, nodePort);
 		if (connection == NULL)
 		{
 			failedPlacementList = lappend(failedPlacementList, taskPlacement);
 			continue;
 		}
 
-		result = PQexec(connection, task->queryString);
-		if (PQresultStatus(result) != PGRES_COMMAND_OK)
+		queryOK = SendQueryInSingleRowMode(connection, task->queryString);
+		if (!queryOK)
 		{
-			ReportRemoteError(connection, result);
-			PQclear(result);
-
+			PurgeConnection(connection);
 			failedPlacementList = lappend(failedPlacementList, taskPlacement);
 			continue;
 		}
 
-		currentAffectedTupleString = PQcmdTuples(result);
-		currentAffectedTupleCount = pg_atoi(currentAffectedTupleString, sizeof(int32), 0);
-
-		if ((affectedTupleCount == -1) ||
-			(affectedTupleCount == currentAffectedTupleCount))
+		/*
+		 * If caller is interested, store query results the first time
+		 * through. The output of the query's execution on other shards is
+		 * discarded if we run there (because it's a modification query).
+		 */
+		if (!gotResults && expectResults)
```

Comment:
I think the word `results` a bit confusing. Instead of `gotResults` and `expectResults`, using `receiviedTuples` and `expectTuples` may look better. What do you think?